{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63235cd0-72f1-469c-a1a3-4b52f498f47f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Answer 1\n",
    "#we use exception class while using custom exception because we want to use the same exception class for all the exceptions.\n",
    "\n",
    "#Answer 2\n",
    "def print_exception_hierarchy(exception_class, indent=\"\"):\n",
    "    print(indent + exception_class.__name__)\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + \"  \")\n",
    "\n",
    "print_exception_hierarchy(BaseException)\n",
    "\n",
    "#Answer 3\n",
    "#Errors in arithmetic error class are: OverflowError, ZeroDivisionError, FloatingPointError\n",
    "\n",
    "try:\n",
    "    a = int(input(\"Enter a number: \"))\n",
    "    b = int(input(\"Enter a number: \"))\n",
    "    c = a/b\n",
    "    print(c)\n",
    "except ArithmeticError as e:\n",
    "    print(\"Arithmetic error\",e)\n",
    "\n",
    "\n",
    "\n",
    "try:\n",
    "    a1= 100 ** 10000000\n",
    "    b1= a1 * a1\n",
    "    print(b1)\n",
    "except ArithmeticError as e:\n",
    "    print(\"Overflow error\",e)\n",
    "\n",
    "\n",
    "#Answer 4\n",
    "def lookup_example(data, key):\n",
    "    try:\n",
    "        value = data[key]\n",
    "        print(\"Value found:\", value)\n",
    "    except LookupError:\n",
    "        print(\"LookupError: Key not found or index out of range\")\n",
    "\n",
    "data_dict = {\"a\": 1, \"b\": 2, \"c\": 3}\n",
    "data_list = [10, 20, 30, 40, 50]\n",
    "data_string = \"Hello, world!\"\n",
    "\n",
    "lookup_example(data_dict, \"b\")  \n",
    "lookup_example(data_list, 2)     \n",
    "lookup_example(data_string, 12)  \n",
    "\n",
    "#Answer 6\n",
    "#Best practices for exception handling:\n",
    "#1. Use specific exception instead of general exception.\n",
    "#2. Use try-except instead of if-else.\n",
    "#3. Use finally block to close the resources.\n",
    "#4. Use exception chaining to preserve the original exception.\n",
    "#5. Use exception handling for programmer errors.\n",
    "#6. Use exception handling for user input errors.\n",
    "#7. Use exception handling for third party libraries.\n",
    "#8. Use exception handling for system errors.\n",
    "#9. Use exception handling for expected errors.\n",
    "#10. Use exception handling for unexpected errors.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
